framework/src/main/java/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -538,6 +544,31 @@ private Range calculateLengthRangeAddition(Range leftLengths, Range rightLengths
         return leftLengths.plus(rightLengths).intersect(Range.INT_EVERYTHING);
     }
 
+    /**
+     * This method is only used to check for nullable nodes in string concatenations.
+     *
+     * @return false if the node can't take null value in a String concatenation or true otherwise.
+     *     There is an exception for StringConcatenationNode, as there is no need no check its
+     *     operands individually.
+     */
+    private boolean isNullable(Node node) {
+        if (node instanceof StringLiteralNode) {
+            return false;
+        } else if (node instanceof StringConversionNode) {
+            if (((StringConversionNode) node).getOperand().getType().getKind().isPrimitive()) {
+                return false;
+            }
+        } else if (node instanceof StringConcatenateNode) {
+            // no need to recurse because this method is called for every operand.
```

Comment:
I don't understand this comment.  There's no recursion.

Reply:
Sorry for being imprecise here. What I meant is:

In String concatenation `"hello" + 12 + Integer.MAX_VALUE`, the first operand is `"hello" + 12 and the second operand is `Integer.MAX_VALUE`. I wanted to say that if a `StringConcatenationNode` enters the method `isNullable` (like the first operand in my example is), then it is not necessary to check the operands of this `StringConcatenationNode` because the method `is Nullable` will eventually be called for the operands of String concatenation `"hello" + 12` too, since it is a String concatenation like any other.

For these simple examples it is not a problem if `isNullable` is called recursively for the operands of a `StringConcatenationNode`, but in some test cases there are some long String concatenations (see `framework/tests/StringConcats.java` line 7) in which these checks would be performed many times for no reason.

Hopefully this explains it. I deleted the comment while thinking of something more descriptive to write there.

Reply:
Thanks for the explanation.  There's no need to call isNullable recursively because a StringConcatenateNode can never be `null`.